File: ObscuritasMediaManager.Frontend/advanced-components/music-filter/music-filter-options.js
import { CheckboxState } from '../../data/enumerations/checkbox-state.js';
import { FilterEntry } from '../../data/filter-entry.js';
import { session } from '../../data/session.js';
import {
    Instrumentation,
    InstrumentType,
    Mood,
    MusicGenre,
    Nation,
    Participants,
} from '../../obscuritas-media-manager-backend-client.js';

export class MusicFilterOptions {
    /** @type {string} */ search = '';
    /** @type {CheckboxState} */ showComplete = CheckboxState.Ignore;
    /** @type {CheckboxState} */ showPlaylists = CheckboxState.Ignore;
    /** @type {CheckboxState} */ showDeleted = CheckboxState.Forbid;

    languages = new FilterEntry(Object.values(Nation), CheckboxState.Require);
    ratings = new FilterEntry(['1', '2', '3', '4', '5'], CheckboxState.Require);
    nations = new FilterEntry(Object.values(Nation), CheckboxState.Require);
    instrumentTypes = new FilterEntry(Object.values(InstrumentType));
    instruments = new FilterEntry(
        session.instruments.current().map((x) => x.name),
        CheckboxState.Ignore
    );
    moods = new FilterEntry(Object.values(Mood), CheckboxState.Ignore);
    genres = new FilterEntry(Object.values(MusicGenre), CheckboxState.Ignore);
    instrumentations = new FilterEntry(Object.values(Instrumentation), CheckboxState.Ignore);
    participants = new FilterEntry(Object.values(Participants), CheckboxState.Require);

    /**
     * @param {string} text
     */
    static fromJSON(text) {
        var object = JSON.parse(text);
        for (var key in object) if (object[key]?.states) Object.setPrototypeOf(object[key], FilterEntry.prototype);
        return object;
    }
}

File: ObscuritasMediaManager.Client/ObscuritasMediaManager.Client/BusinessComponents/MusicFilter/MusicFilterOptions.cs
using ObscuritasMediaManager.Backend.Data.Music;
using System;
using System.Linq;

namespace ObscuritasMediaManager.Client.BusinessComponents.MusicFilter;

public class MusicFilterOptions
{
    public FilterEntry<MusicGenre> genres { get; set; } = new(Enum.GetValues<MusicGenre>(), CheckboxState.Ignore);
    public FilterEntry<Instrumentation> instrumentations
    {
        get;
        set;
    } 
        = 
        new(Enum.GetValues<Instrumentation>(), CheckboxState.Ignore);
    public FilterEntry<string> instruments;
    public FilterEntry<InstrumentType> instrumentTypes { get; set; } = new(Enum.GetValues<InstrumentType>());
    public FilterEntry<Nation> languages { get; set; } = new(Enum.GetValues<Nation>(), CheckboxState.Require);
    public FilterEntry<Mood> moods { get; set; } = new(Enum.GetValues<Mood>(), CheckboxState.Ignore);
    public FilterEntry<Nation> nations { get; set; } = new(Enum.GetValues<Nation>(), CheckboxState.Require);
    public FilterEntry<Participants> participants { get; set; } = new(Enum.GetValues<Participants>(), CheckboxState.Require);
    public FilterEntry<int> ratings { get; set; } = new(new [] { 1, 2, 3, 4, 5 }, CheckboxState.Require);
    public string search { get; set; } = string.Empty;
    public CheckboxState showComplete { get; set; } = CheckboxState.Ignore;
    public CheckboxState showDeleted { get; set; } = CheckboxState.Forbid;
    public CheckboxState showPlaylists { get; set; } = CheckboxState.Ignore;

    public MusicFilterOptions(IEnumerable<string> instrumentNames)
    {
        instruments = new(instrumentNames, CheckboxState.Ignore);
    }
}
